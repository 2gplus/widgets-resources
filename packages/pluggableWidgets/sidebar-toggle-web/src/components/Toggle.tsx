import { createElement, CSSProperties, HTMLAttributes, ReactElement } from "react";
import classNames from "classnames";
import { WebIcon } from "mendix";
import { Icon } from "@mendix/piw-utils-internal/components/web";

interface ToggleProps {
    caption?: string;
    className: string;
    icon?: WebIcon;
    tabIndex?: number;
    tooltip?: string;
    render: "button" | "link";
    role?: string;
    style?: CSSProperties;
}

export function Toggle(props: ToggleProps): ReactElement {
    const commonProps = {
        tabIndex: props.tabIndex,
        title: props.tooltip,
        "aria-label": props.tooltip,
        style: props.style
    };

    const icon = <Icon icon={props.icon} className="widget-sidebar-toggle-icon" fallback={<span />} />;

    const prepareRole = (text?: string): string | undefined => text?.replace(/_/g, "-");

    const action = (): void => {
        document.dispatchEvent(new CustomEvent("toggleSidebar"));
    };

    const events: Partial<HTMLAttributes<HTMLAnchorElement | HTMLButtonElement>> = {
        onClick: e => {
            e.preventDefault();
            e.stopPropagation();
            action();
        },
        onKeyDown: e => {
            e.preventDefault();
            e.stopPropagation();

            if (e.key === " " || e.key === "Enter") {
                action();
            }
        }
    };

    if (props.render === "link") {
        return (
            <a
                {...commonProps}
                role={prepareRole(props.role)}
                className={classNames("widget-sidebar-toggle", "mx-link", props.className)}
                {...events}
            >
                {icon}
                {props.caption}
            </a>
        );
    }
    return (
        <button
            {...commonProps}
            className={classNames("widget-sidebar-toggle", "btn mx-button toggle-btn", props.className)}
            {...events}
        >
            {icon}
            {props.caption}
        </button>
    );
}
