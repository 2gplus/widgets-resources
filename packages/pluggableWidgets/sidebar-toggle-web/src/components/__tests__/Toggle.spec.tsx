import { createElement } from "react";
import { Toggle } from "../Toggle";
import { mount, render } from "enzyme";

describe("Sidebar toggle", () => {
    describe("rendering", () => {
        it("renders correctly as a link", () => {
            const toggle = render(<Toggle className="" render="link" role="button" />);

            expect(toggle).toMatchSnapshot();
        });

        it("renders correctly as a button", () => {
            const toggle = render(<Toggle className="" render="button" />);

            expect(toggle).toMatchSnapshot();
        });

        it("renders correctly with an icon", () => {
            const toggle = render(
                <Toggle className="" render="button" icon={{ type: "glyph", iconClass: "hamburger" }} />
            );

            expect(toggle).toMatchSnapshot();
        });

        it("renders correctly with a tooltip and caption", () => {
            const toggle = render(<Toggle className="" render="button" tooltip="this is a toggle" caption="menu" />);

            expect(toggle).toMatchSnapshot();
        });
    });

    describe("behaviors", () => {
        it("calls correct event when pressing link", () => {
            const fn = jest.fn();
            const toggle = mount(<Toggle className="" render="link" role="button" />);

            document.addEventListener("toggleSidebar", fn);

            toggle.simulate("click");

            expect(fn).toBeCalled();
        });

        it("calls correct event when pressing button", () => {
            const fn = jest.fn();
            const toggle = mount(<Toggle className="" render="button" />);

            document.addEventListener("toggleSidebar", fn);

            toggle.simulate("click");

            expect(fn).toBeCalled();
        });

        it("calls correct event when pressing enter over a link", () => {
            const fn = jest.fn();
            const toggle = mount(<Toggle className="" render="link" role="button" />);

            document.addEventListener("toggleSidebar", fn);

            toggle.simulate("keydown", {
                preventDefault: jest.fn(),
                stopPropagation: jest.fn(),
                key: "Enter"
            });

            expect(fn).toBeCalled();
        });

        it("calls correct event when pressing enter over a button", () => {
            const fn = jest.fn();
            const toggle = mount(<Toggle className="" render="button" />);

            document.addEventListener("toggleSidebar", fn);

            toggle.simulate("keydown", {
                preventDefault: jest.fn(),
                stopPropagation: jest.fn(),
                key: "Enter"
            });

            expect(fn).toBeCalled();
        });
    });

    describe("structure", () => {
        it("contains custom classes", () => {
            const toggle = mount(<Toggle className="custom-class" render="button" />);

            expect(toggle.find("button").prop("className")).toContain("custom-class");
        });

        it("applies correct role to links", () => {
            const toggle = mount(<Toggle className="" render="link" role="my_underscored_role" />);

            expect(toggle.find("a").prop("role")).toContain("my-underscored-role");
        });
    });
});
