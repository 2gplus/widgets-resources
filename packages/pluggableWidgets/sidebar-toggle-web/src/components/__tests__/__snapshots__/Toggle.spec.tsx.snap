// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar toggle rendering renders correctly as a button 1`] = `
<button
  class="widget-sidebar-toggle btn mx-button toggle-btn"
/>
`;

exports[`Sidebar toggle rendering renders correctly as a link 1`] = `
<a
  class="widget-sidebar-toggle mx-link"
  role="button"
/>
`;

exports[`Sidebar toggle rendering renders correctly with a tooltip and caption 1`] = `
<button
  aria-label="this is a toggle"
  class="widget-sidebar-toggle btn mx-button toggle-btn"
  title="this is a toggle"
>
  menu
</button>
`;

exports[`Sidebar toggle rendering renders correctly with an icon 1`] = `
<button
  class="widget-sidebar-toggle btn mx-button toggle-btn"
>
  <span
    aria-hidden="true"
    class="glyphicon widget-sidebar-toggle-icon hamburger"
  />
</button>
`;
